### PR TITLE
[E & A] Implement colormap configurability

### PR DESCRIPTION
**Pull request by dzole0311**
**Related Ticket:** https://github.com/NASA-IMPACT/veda-ui/issues/994

### Description of Changes

Implements the color map configuration of E&A layers.

Changes:

- Adds curated list of supported color maps taken from the back-end, split into diverging and sequential groups
- Supports other valid color map names from the list of color maps: https://openveda.cloud/api/raster/colorMaps
- Adds support for reversing the colors
- Adds reset all button which currently only resets the "Reverse" option, but will also apply to the "Rescale" functionality which is to be added next
- Added a loading skeleton to the colormap in the timeline sidebar when a layer is added. This prevents a visual flash because it might take a second to fetch the color map from the back-end (cc @faustoperez )

The precedence of the color maps are as follows:

1. Start with what is configured in the sourceParams of the dataset
2. If it's not defined, check for the dashboard render configuration coming from STAC endpoint
3. If still undefined, check for asset-specific renders using the sourceParams' assets. This is a special edge case for some datasets that have multiple assets (multiple assets under one collection item)
4. If all else fails, default to `viridis`

Please let me know your thoughts on that or if you spot something strange

### Notes & Questions About Changes

- I additionally pushed a shell script that I used to build the `colorMaps.ts` file, I hope that's okay. It fetches all the `rgba` values for each supported colorMap from titiler. We could extend it and integrate it within our Github workflow in the future, so that the colors are fetched on each new release (although it's unlikely they will change that often)
 
### Validation / Testing

- Verify that the color map list is correctly split into diverging and sequential groups
- Check that reversing the color map works as expected and toggles correctly
- Test the "Reset All" button to ensure that it resets the "Reverse" option and works as intended
- Check that the precedence of the color maps is followed in the correct order:

1. Start with sourceParams of the dataset
2. Fallback to dashboard render configuration from the STAC endpoint
3. If undefined, check for asset-specific renders using sourceParams' assets
4. Finally, default to `viridis` if all else fails

- Verify that the changes do not break any existing functionality related to rendering of the tiles on the map (selected color map must match the tile colors)
- Test across different datasets with both single and multiple assets to confirm edge cases are handled as described

### CODE REVIEW

**Comment by j08lue** [comment]
A few questions:

1. Vector datasets like our demo dataset "EIS Fire" should probably not show a colorbar at all, right? 

   image

2. Should the "Reset All" button not also reset the colormap to the default?
3. I had a design-related question regarding the units, too. https://github.com/NASA-IMPACT/veda-ui/issues/994#issuecomment-2304312833

**Comment by faustoperez** [comment]
> Added a loading skeleton to the colormap in the timeline sidebar when a layer is added. This prevents a visual flash because it might take a second to fetch the color map from the back-end (cc @faustoperez )

Sounds great @dzole0311 !

**Review by hanbyul-here — CHANGES_REQUESTED** [review]
This needs to be changed if we want to give a priority to manual sourceParams: https://github.com/NASA-IMPACT/veda-ui/pull/1117/files#diff-fc839f8286ca1783e679e0fcddd8a3da452570ebbe9f433e16fda7003b91687eR376 (but hopefully we can move on from manual configurations soon 😬 !)

**Comment by hanbyul-here** [comment]
One more - do you see any reason not to apply dynamic colormap to other types of raster layers? (cmr and zarr)? - We can separate the issues for these layers since these are not being used on production anyway, but I was wondering if there is any specific reason!

File: app/scripts/components/common/map/style-generators/raster-timeseries.tsx [hunk 1]
```diff
@@ -36,6 +36,20 @@ import {
 // Whether or not to print the request logs.
 const LOG = true;
 
+export interface RasterTimeseriesProps extends BaseGeneratorParams {
```
[the remark this answers is not in the document]
Comment (dzole0311) [reply app/scripts/components/common/map/style-generators/raster-timeseries.tsx]: Whoops, I had a merge conflict and this ended up here. Moved it back in the `types` where it was.

File: app/scripts/components/exploration/components/datasets/colormap-options.tsx [hunk 2]
```diff
@@ -0,0 +1,195 @@
+import React, { useEffect, useState } from 'react';
+import styled from 'styled-components';
+import { Button, Icon } from "@trussworks/react-uswds";
+import { CollecticonDrop } from '@devseed-ui/collecticons';
+import { sequentialColorMaps, divergingColorMaps, restColorMaps } from './colorMaps';
+
+import './colormap-options.scss';
+
+const Container = styled.div`
+  width: 383px;
```
[the remark this answers is not in the document]
Comment (dzole0311) [reply app/scripts/components/exploration/components/datasets/colormap-options.tsx]: I tried, but I can't find one that fits this use case.

File: app/scripts/components/exploration/components/datasets/colormap-options.tsx [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+  const initialColorMap = normalizeColorMap(colorMap);
+
+  const [isReversed, setIsReversed] = useState(initialIsReversed);
+  const [selectedColorMap, setSelectedColorMap] = useState(initialColorMap);
+
+  const colormapType = classifyColormap(selectedColorMap);
+  const [customColorMap, setCustomColorMap] = useState<string | null>(null);
+
+  useEffect(() => {
+    if (colormapType === 'sequential' && !CURATED_SEQUENTIAL_COLORMAPS.includes(selectedColorMap)) {
```
[the remark this answers is not in the document]
Comment (dzole0311) [reply app/scripts/components/exploration/components/datasets/colormap-options.tsx:64]: Yes, exactly. We want to support other valid titiler colormap names (the ones currently returned by the titiler `colorMap` endpoint) and ideally map them to `sequential` or `diverging` groups.

File: app/scripts/components/common/map/layer-legend.tsx [hunk 4]
```diff
@@ -300,25 +302,47 @@ export function LayerCategoricalGraphic(props: LayerLegendCategorical) {
   );
 }
 
-export function LayerGradientGraphic(props: LayerLegendGradient) {
-  const { stops, min, max, unit } = props;
-
+export const LayerGradientGraphic = (props: LayerLegendGradient) => {
```
[the remark this answers is not in the document]
Comment (dzole0311) [reply app/scripts/components/common/map/layer-legend.tsx:304]: They are already in the source code and outside of the faux module (at least the `LayerLegendGradient` type). Am I missing something? 🤔

File: app/scripts/components/common/map/style-generators/raster-timeseries.tsx [hunk 5]
```diff
@@ -358,7 +374,8 @@ export function RasterTimeseries(props: RasterTimeseriesProps) {
         const tileParams = qs.stringify(
           {
             assets: 'cog_default',
-            ...(sourceParams ?? {})
+            ...(sourceParams ?? {}),
+            colormap_name: colorMap
```
[the remark this answers is not in the document]
Comment (dzole0311) [reply app/scripts/components/common/map/style-generators/raster-timeseries.tsx]: We need [this](https://github.com/NASA-IMPACT/veda-ui/pull/1117/files#diff-b015ae8d170bf7762b2b8ad8c59673574d21cced66fccabdf7cac68e7d30b864R27) in order to check if the `colorMap` is part of the url params. Did you mean we could move it one level up or completely get rid of it?

File: app/scripts/components/exploration/components/datasets/data-layer-card.tsx [hunk 6]
```diff
@@ -158,19 +175,41 @@ export default function DataLayerCard(props: CardProps) {
...
-            type='gradient'
-            stops={datasetLegend.stops}
-            min={datasetLegend.min}
-            max={datasetLegend.max}
-          />
+        {/* Show a loading skeleton when the color map is not categorical and the dataset
+        status is 'loading'. This is because we color map sometimes might come from the titiler
+        which could introduce a visual flash when going from the 'default' color map to the one
+        configured in titiler */}
+        {!isColorMapCategorical && dataset.status === 'loading'  && datasetLegend?.type === 'gradient' && <LoadingSkeleton />}
```
[the remark this answers is not in the document]
Comment (dzole0311) [reply app/scripts/components/exploration/components/datasets/data-layer-card.tsx]: I removed the `isColorMapCategorical` completely as it was not considering vector layer cases. The preloading skeleton should be fixed as well.

**Comment by dzole0311** [comment]
Lots of good comments @hanbyul-here, thanks!

> One more - do you see any reason not to apply dynamic colormap to other types of raster layers? (cmr and zarr)? - We can separate the issues for these layers since these are not being used on production anyway, but I was wondering if there is any specific reason!

It somehow slipped my mind to pass the `colormap` to the `RasterPaintLayer` as well. Should work now for CMR and Zarr layers.

File: app/scripts/components/exploration/hooks/use-stac-metadata-datasets.ts [hunk 7]
```diff
@@ -43,12 +52,47 @@ function reconcileQueryDataWithDataset(
...
+        renderParams = queryData.data.renders.dashboard;
+      }
+
+      if (
+        !renderParams &&
+        queryData.data.renders?.[base.data.sourceParams?.assets]
+      ) {
+        renderParams = queryData.data.renders[base.data.sourceParams?.assets];
+
+        // The backend sometimes sends `renderParams.rescale` as a nested array, e.g., [[min, max]],
```
[the remark this answers is not in the document]
Comment (dzole0311) [reply app/scripts/components/exploration/hooks/use-stac-metadata-datasets.ts]: I realized I may not have fully clarified why flattening is necessary here (please feel free to correct me if I'm off). When Parcel parses the MDX rescale values, they end up in the format [0, 0.1, 0.01, 0.2]. While our UI handles this well (though I haven't encountered multiple rescale values in the GHG Center instance or across our mocked data), the back-end returns rescale values in a [nested format](https://github.com/developmentseed/titiler/blob/dd773230e1d906625f3cdf12933dfe699e952919/src/titiler/core/titiler/core/dependencies.py#L445), like [[0, 0.1], [0.01, 0.2]]. To make sure we safe guard against both formats so that the UI correctly gets only the min,max values it needs, I believe we need to flatten the `rescale` values from the back-end as a precaution.

File: app/scripts/components/common/map/layer-legend.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (hanbyul-here) [reply app/scripts/components/common/map/layer-legend.tsx:304]: Oh I am sorry, I just assumed from parcel-resolver-veda change is for this one and the change in the file is not related to this.

File: docs/content/frontmatter/layer.md [hunk 8]
```diff
@@ -34,85 +34,93 @@ info: Info
...
-  `string`  
-  The colormap to use for the layer. One of https://cogeotiff.github.io/rio-tiler/colormap/#default-rio-tilers-colormaps
+**sourceParams**
+`object`
+Parameters to be appended to the tile source. These values will be used as query string parameters and their priority is determined based on the following precedence:
+
+1. **Start with the configuration in the `sourceParams` of the dataset.**
+2. **If not defined, fallback to the dashboard render configuration from the STAC endpoint.**
+3. **If still undefined, check for asset-specific renders using the `sourceParams`' `assets` property.**
+4. **Finally, default to 'viridis' if all else fails.**
```
Comment (hanbyul-here) [inline docs/content/frontmatter/layer.md:83]:
```suggestion
4. **Finally, default to `{colormap_name: 'viridis'}` if all else fails.**
```

**Review by aboydnw — COMMENTED** [review]
I think that somehow leading users back to the visualization recommended/configured by the science team will be important to stakeholders. Whether that's the reset button also resetting the colormap selection or providing some sort of visual cue which colormap is the default.

I see in the netlify preview that some of the datasets start with "Default," and some didn't. I assume this has to do with the logic you were talking about @dzole0311 to decide which colormap is the default. Is there any way to ensure that whatever is decided to be the default comes with this tag?

Also, another question, do we think users will want to see both Default AND the colormap name? cc @faustoperez 

Example w/o default tag:
image

There is also an edge case that I discovered. If it's an easy fix, maybe worth addressing. Otherwise we can make a ticket and do it at some point in the future.
Steps to reproduce:
1. Open exploration tab and select layers
2. Change the colormap of one of the layers to something other than the Default
3. Notice that if you close and reopen the colormap bar, the Default tag is still there
4. Refresh the page and notice that the change in colormap is maintained (I believe this is desirable, so people can share views with others via URL)
5. Open the colormap bar and notice that the Default tag is no longer there

File: app/scripts/components/exploration/hooks/use-stac-metadata-datasets.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (hanbyul-here) [reply app/scripts/components/exploration/hooks/use-stac-metadata-datasets.ts]: Yeah, it makes sense to follow the existing pattern - But I think our MDX configuration is outdated, and we should change the system to accommodate the multiple bands - let's ticket this separately.

File: app/scripts/components/exploration/components/datasets/colormap-options.tsx [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+  const colorKeys = Object.keys(colormapData);
+  const colors = colorKeys.map((key) => {
+    const [r, g, b, a] = colormapData[key];
+    return `rgba(${r}, ${g}, ${b}, ${a})`;
+  });
+
+  return isReversed ? colors.reduceRight((acc, color) => [...acc, color], []) : colors;
+};
+
+export function ColormapOptions({ colorMap, setColorMap}: ColormapOptionsProps) {
```
Comment (sandrahoang686) [inline app/scripts/components/exploration/components/datasets/colormap-options.tsx:53]: Curious, with Jotai state management, cant we access the same state from the same atom? so instead of passing `colorMap` state and setter method, can we just pass in the `datasetAtom` and then do `const [colorMap, setColorMap] = useTimelineDatasetColormap(datasetAtom);` at this level to avoid prop drilling since `datasetAtom` was already being previously being passed into `DataLayerCard`?

File: app/scripts/components/common/map/layer-legend.tsx [hunk 10]
```diff
@@ -340,4 +345,40 @@ export function LayerGradientGraphic(props: LayerLegendGradient) {
...
+    if (Array.isArray(value) && value.length === 4) {
+      return `rgba(${value.join(',')})`;
+    } else {
+      return `rgba(0, 0, 0, 1)`;
+    }
+  });
+
+  const processedStops = colormap.isReversed
+  ? stops.reduceRight((acc, stop) => [...acc, stop], [])
+  : stops;
```
[the remark this answers is not in the document]
Comment (dzole0311) [reply app/scripts/components/common/map/layer-legend.tsx:368]: `reverse` mutates the array in place, while the `reduceRight` approach here creates a new array. In the case of the first one, TypeScript will complain.

**Comment by dzole0311** [comment]
As discussed yesterday, I removed the "Reset All" and "Default" label for now.

> I see in the netlify preview that some of the datasets start with "Default," and some didn't. I assume this has to do with the logic you were talking about @dzole0311 to decide which colormap is the default. Is there any way to ensure that whatever is decided to be the default comes with this tag?

@aboydnw This and the other issue you spotted are partially related (ticketed here: https://github.com/NASA-IMPACT/veda-ui/issues/1130) and will be handled together within one fix.

**Review by hanbyul-here — APPROVED** [review]
🙇 Thanks for amazing work!

**Comment by faustoperez** [comment]
> Also, another question, do we think users will want to see both Default AND the colormap name? cc @faustoperez

"Default" here refers to the colormap set by the science team. When designing this, my understanding was that they often don't name their colormaps or use readily available ones like the Viridis family, hence my use of "default" to mean the original setting.

Perhaps replacing "Default" with "Initial" would make this clearer?

Also, if we can identify whether the science team's colormap is one already listed, we could set it as the default or highlight it in the colormap panel to avoid duplicates 🤷‍♂️
